src/pybind/mgr/crash/module.py

```diff
@@ -0,0 +1,248 @@
+        key = 'crash/%s' % crashid
+        if self.get_store(key) is None:
+            return errno.EINVAL, '', 'crash rm: id %s not present' % crashid
+        self.set_store(key, None)       # removes key
+        return 0, '', ''
+
+    def do_prune(self, cmd, inbuf, testmeta=None):
+        now = datetime.datetime.utcnow()
+
+        keep = cmd['keep']
+        try:
+            keep = int(keep)
+        except ValueError:
+            return errno.EINVAL, '', 'keep argument must be integer'
+
+        keeptime = datetime.timedelta(days=keep)
+
+        # testability
+        if testmeta is not None:
+            iterator = testmeta
+            test_removed = list()
+        else:
+            iterator = self.get_store_prefix('crash/')
+
+        for key, meta in iterator.iteritems():
+            meta = json.loads(meta)
+            stamp = self.time_from_string(meta['timestamp'])
+            if stamp <= now - keeptime:
+                if testmeta:
+                    test_removed.append('rm(%s)' % key)
+                else:
+                    self.do_rm(cmd, inbuf)
+
+        if testmeta:
+            return 0, '\n'.join(test_removed), ''
+
+        return 0, '', ''
+
+    def do_stat(self, cmd, inbuf, testmeta=None):
+        # age in days for reporting, ordered smallest first
+        bins = [1, 3, 7]
+        retlines = list()
+
+        def binstr(bindict):
+            binlines = list()
+            count = len(bindict['idlist'])
+            if count:
+                binlines.append(
+                    '%d older than %s days old:' % (count, bindict['age'])
+                )
+                for crashid in bindict['idlist']:
+                    binlines.append(crashid)
+            return '\n'.join(binlines)
+
+        total = 0
+        now = datetime.datetime.utcnow()
+        for i, age in enumerate(bins):
+            agelimit = now - datetime.timedelta(days=age)
+            bins[i] = {
+                'age': age,
+                'agelimit': agelimit,
+                'idlist': list()
+            }
+
+        # testability
+        if testmeta is not None:
+            iterator = testmeta
+        else:
+            iterator = self.get_store_prefix('crash/')
+
+        for key, meta in iterator.iteritems():
+            total += 1
+            meta = json.loads(meta)
+            stamp = self.time_from_string(meta['timestamp'])
+            crashid = meta['crash_id']
+            for i, bindict in enumerate(bins):
+                if stamp <= bindict['agelimit']:
+                    bindict['idlist'].append(crashid)
+                    # don't count this one again
+                    continue
+
+        retlines.append('%d crashes recorded' % total)
+
+        for bindict in bins:
+            retlines.append(binstr(bindict))
+        return 0, '\n'.join(retlines), ''
+
+    def do_self_test(self, cmd, inbuf):
+        # test time conversion
+        timestr = '2018-06-22 20:35:38.058818Z'
```

Comment:
You need to add a case in test_module_selftest.py to get this command called during testing.

Although many modules just have the call to their selftest method, that's usually because they have onerous external dependencies like influxdb/zabbix servers that would be required for more realistic testing.  This module should be fairly easy to test though -- you could create a new test_crash.py in qa/tasks/mgr, with a couple of test cases that drive all this externally with real commands.

Reply:
if I do that, I can remove the selftest for prune and stat altogether, which is probably better.